src/pointer.rs

```diff
@@ -0,0 +1,88 @@
+//! Abstractions over raw pointers.
+
+mod ptr;
+
+pub use maybe_aligned::MaybeAligned;
+pub use ptr::{invariant, Ptr};
+
+/// A shorthand for a maybe-valid, maybe-aligned reference. Used as the argument
+/// to [`TryFromBytes::is_bit_valid`][crate::TryFromBytes::is_bit_valid].
+pub type MaybeValid<'a, T> =
+    Ptr<'a, T, { invariant::Shared }, { invariant::AnyAlignment }, { invariant::AsInitialized }>;
+
+/// A semi-user-facing wrapper type representing a maybe-aligned reference, for
+/// use in [`TryFromBytes::is_bit_valid`][crate::TryFromBytes::is_bit_valid].
+mod maybe_aligned {
+    use super::{ptr::invariant, Ptr};
+    use crate::Unaligned;
+    use core::fmt::{Debug, Formatter};
+
+    /// A reference to a validly-initialized `T` of lifetime `'a` that might
+    /// *not* be validly aligned for `T`.
+    //  NOTE: Although this type is called `MaybeAligned`, that name is only
+    //  appropriate in its sole place of use: `TryFromBytes::is_bit_valid`.
+    //  There, the preconditions on the function's argument ensure that
+    #[doc(hidden)]
+    #[allow(unreachable_pub)] // false positive?
+    pub struct MaybeAligned<'a, T: ?Sized> {
+        ptr: Ptr<'a, T, { invariant::Shared }, { invariant::AnyAlignment }, { invariant::Valid }>,
+    }
+
+    impl<'a, T> MaybeAligned<'a, T>
+    where
+        T: 'a + ?Sized,
+    {
+        /// Constructs a `MaybeAligned` from a [`Ptr`].
+        ///
+        /// The `Ptr`'s referent must be validly-initialized for `T`.
+        #[doc(hidden)]
+        #[inline]
+        pub const fn from_ptr(
+            ptr: Ptr<
+                'a,
+                T,
+                { invariant::Shared },
+                { invariant::AnyAlignment },
+                { invariant::Valid },
+            >,
+        ) -> Self {
+            Self { ptr }
+        }
+
+        /// Reads the value from `MaybeAligned`.
+        ///
+        /// This is only available if `T` is [`Copy`].
+        #[inline]
+        pub fn read_unaligned(self) -> T
+        where
+            T: Copy,
```

Comment:
We might want to figure out a way to relax this bound. One of the use cases for `Ptr` is to go from `[u8]` to `T`, and we support `T: FromBytes` even when `T: !Copy`. In principle, we should support reading a `T` from a byte slice even if `T: !Copy`. I grant that, in this particular case, it would be unsound to omit `T: Copy` because the `Ptr` could have been constructed from a `&Box<...>` or something.

Reply:
Maybe let's open an issue to track having `Ptr` store the "original" type of its data? Probably not worth holding up this PR for that, and I suspect there are a lot of subtleties.